File: app/src/main/java/com/miandroidchallenge/ucoppp/miandroidchallenge/ui/deliverylistfragment/DeliveryFragmentFactoryViewModel.java
package com.miandroidchallenge.ucoppp.miandroidchallenge.ui.deliverylistfragment;

import android.app.Application;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.miandroidchallenge.ucoppp.miandroidchallenge.ui.deliverylistfragment.interfaces.OnDeliveryChange;

import javax.inject.Inject;

public class DeliveryFragmentFactoryViewModel implements ViewModelProvider.Factory {

    private Application application;

    private OnDeliveryChange onDeliveryChange;

    @Inject
    public DeliveryFragmentFactoryViewModel(Application application, OnDeliveryChange onDeliveryChange) {
        this.application = application;
        this.onDeliveryChange = onDeliveryChange;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(DeliveryFragmentViewModel.class)) {
            return (T) new DeliveryFragmentViewModel(application, onDeliveryChange);
        }
        throw new IllegalArgumentException("Unknown" + modelClass.getName() + "class");
    }
}

File: app/src/main/java/com/miandroidchallenge/ucoppp/miandroidchallenge/ui/deliverylistfragment/DeliveryFragmentFactoryViewModel.kt
package com.miandroidchallenge.ucoppp.miandroidchallenge.ui.deliverylistfragment

import android.app.Application
import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import com.miandroidchallenge.ucoppp.miandroidchallenge.ui.deliverylistfragment.interfaces.OnDeliveriesChange
import javax.inject.Inject

class DeliveryFragmentFactoryViewModel @Inject constructor(val application: Application,
                                                           private val onDeliveriesChange: OnDeliveriesChange
) : ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(DeliveryFragmentViewModel::class.java)) {
            return DeliveryFragmentViewModel(application, onDeliveriesChange) as T
        }
        throw IllegalArgumentException("Unknown ${modelClass.name} class")
    }
}